File: copy_dir_struct/copy_dir_struct.pl
#!/usr/bin/perl
# By Raphael Champeimont, public domain

# This programs copies a directory structure.

use strict;
use warnings;

sub usage {
  print "Usage: $0 source_dir dest_dir\n";
  exit 0;
}

sub dodir {
  my ($sdir, $path, $ddir) = @_;
  die if (not defined $ddir);

  # list dir contents
  opendir(DIR, "$sdir$path") || die "can't opendir \"$ddir$path\": $!";
  my @subdirs = readdir DIR;
  closedir DIR;

  # make dir copy
  print "mkdir \"$ddir$path\"\n";
  if (system "mkdir", "-p", "$ddir$path") {
    die $!;
  }

  # go in subdirs
  foreach my $dir (@subdirs) {
    next if ($dir eq "." or $dir eq "..");
    next if not -d "$sdir$path/$dir";
    dodir($sdir, "$path/$dir", $ddir);
  }
}

sub main {

  if (not defined $ARGV[1]) {
    usage();
  }

  my $sdir = $ARGV[0];
  my $ddir = $ARGV[1];

  print "Source: $sdir\nDestination: $ddir\n";

  if (not -d $sdir) {
    die "$sdir is not a directory";
  }

  dodir($sdir, "", $ddir);

}

main();

File: copy_dir_struct/copy_dir_struct.py
#!/usr/bin/python
# By Raphael Champeimont, public domain

# This programs copies a directory structure,
# ie is a sort of cp -R without the files.
# Its prurpose is to copy a directory tree to another
# directory (which is mkdired if does not exist).
# The destination directory may be under the source
# dir if and only if it is a direct subdir.

import os;
import sys;



def usage():
  print 'Usage: %s source_dir dest_dir' % sys.argv[0]
  sys.exit(0)



def dodir(sdir, path, ddir):
  # List dir contents, we use "list()" because we want
  # the list of subdirs *before* the new dir is created.
  # This way the program can be used to copy a directory
  # in a direct subdirectory of itself.
  subdirs = list(os.listdir(sdir + path))

  # make dir copy
  print 'mkdir "' + ddir + path + '"'
  try:
    os.makedirs(ddir + path)
  except OSError, e:
    print 'Warning: ' + e.strerror

  # go in subdirs
  for dir in subdirs:
    if not os.path.isdir(sdir + path + '/' + dir):
      continue;
    dodir(sdir, path + '/' + dir, ddir)



def main():
  if len(sys.argv) < 3:
    usage()

  sdir = sys.argv[1]
  ddir = sys.argv[2]

  print "Source: %s\nDestination: %s" % (sdir, ddir)

  if not os.path.isdir(sdir):
    print sdir + " is not a directory"
    sys.exit(1);

  dodir(sdir, '', ddir)



main()
